crypto/evp/pmeth_lib.c

```diff
@@ -112,10 +114,35 @@ const EVP_PKEY_METHOD *EVP_PKEY_meth_find(int type)
     return (**ret)();
 }
 
-static EVP_PKEY_CTX *int_ctx_new(OPENSSL_CTX *libctx,
-                                 EVP_PKEY *pkey, ENGINE *e,
-                                 const char *name, const char *propquery,
-                                 int id)
+EVP_PKEY_METHOD *EVP_PKEY_meth_new(int id, int flags)
+{
+    EVP_PKEY_METHOD *pmeth;
+
+    pmeth = OPENSSL_zalloc(sizeof(*pmeth));
+    if (pmeth == NULL) {
+        EVPerr(EVP_F_EVP_PKEY_METH_NEW, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    pmeth->pkey_id = id;
+    pmeth->flags = flags | EVP_PKEY_FLAG_DYNAMIC;
+    return pmeth;
+}
+#endif
+
+/*- All methods below can also be used in FIPS_MODE */
+
+EVP_PKEY_CTX *EVP_PKEY_CTX_new_provided(OPENSSL_CTX *libctx,
+                                        const char *name,
+                                        const char *propquery)
+{
+    return evp_pkey_ctx_new_int(libctx, NULL, NULL, name, propquery, -1);
+}
+
+EVP_PKEY_CTX *evp_pkey_ctx_new_int(OPENSSL_CTX *libctx,
```

Comment:
I disagree with making this non-static.  I *do* see how you use it in fipsprov.c, but I can also see the use of a public interface that takes a library context and a pkey, and would much have that than surprising internal hackery.

Considering that `EVP_PKEY_CTX_new_provided` is completely new, we're also at liberty to rename it as we see fit.  I could see a set of `EVP_PKEY_CTX_new_from_` functions (`EVP_PKEY_CTX_new_from_name` could replace `EVP_PKEY_CTX_new_provided`, and `EVP_PKEY_CTX_new_from_pkey` could be the name to implement for the fipsprov.c usage).

Reply:
ok..